signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+        self.init_listener.stop()
+        self.subs = []
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+
+class SignalResponder:
+    """
+    A signal management object for tools that wish to respond to event/state signals.
+    Can be used both for listening for signals as well as responding to them. Also
+    allows for locking onto specific tags/publisher_ids. Uses the standard signal
+    response protocol for all published messages.
+    """
+
+    def __init__(self, redis_host: str = "localhost", redis_port: int = 6379) -> None:
+        """
+        Sets exporter object fields and generates unique tool_id.
+        Allows for specification of redis host/port.
+        """
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        self.subscriber.subscribe("event-signal-pubsub")
+        self.tool_id = platform.node() + "-resp"  # ADD UUID HERE?
+        self.locked_id = None
+        self.locked_tag = None
+
+    def _parse_signal(self, signal: Dict) -> Dict:
+        """
+        Validates received signal. Returns payload if valid.
+        Also applies tag/publisher_id filters if added.
+        """
+        try:
+            data = json.loads(signal["data"])
+        except ValueError:
+            return None
+        # FIXME - Replace below line, maybe with dataclasses.fields()?
+        check_set = set(
+            [
+                "publisher_id",
+                "process_name",
+                "event",
+                "runner_host",
+                "sample_no",
+                "tag",
+                "metadata",
+            ]
+        )
+        if set(data.keys()) == check_set or check_set - set(data.keys()) == {
+            "metadata"
+        }:
+            if (not self.locked_id or self.locked_id == data["publisher_id"]) and (
+                not self.locked_tag or self.locked_tag == data["tag"]
+            ):
+                return data
+        return None
+
+    def listen(self):
+        """
+        Yield all legal published signals. If a specific tag/published_id
+        was locked, only signals with those matching values will be yielded.
+        """
+        for item in self.subscriber.listen():
+            data = self._parse_signal(item)
+            if data:
+                yield data
+
+    def respond(self, publisher_id: str, event: str, ras: int = None) -> None:
+        """
+        Publish a legal response to a certain publisher_id's event signal.
+        Also allows for optional ras code to be added on (required for 
+        publisher acknowledgement, but not for initialization response).
+        """
+        response = Response(self.tool_id, publisher_id, event, ras)
+        self.redis.publish("event-signal-response", response.to_json_str())
+
+    def lock_id(self, publisher_id: str) -> None:
```

Comment:
Looks like these are setters for ``locked_id`` and ``lock_tag``, however users can still modify the ``locked_id`` and ``locked_tag`` directly. Python has the concept of properties for this purpose. Can we do one of these:

* Make ``locked_id`` and ``locked_tag`` properties, with a setter that does the validation
* Add underscores in front of ``locked_id`` and ``locked_tag`` to show that they shouldn't be modified by users directly